File: merge_intervals/maximum-cpu-load.js
class Job {
    constructor(start, end, cpu_load) {
        this.start = start;
        this.end = end;
        this.cpu_load = cpu_load;
    }
}

const find_max_cpu_load = (jobs) => {
    jobs.sort((a, b) => a.end - b.end);
    let result = jobs[0].cpu_load;
    let currentLoad = result;
    let back = 0;
    let front = 1;


    while (front < jobs.length) {
        while (back < front) {
            if (jobs[front].start < jobs[back].end) {
                break;
            } else {
                currentLoad -= jobs[back].cpu_load;
                back++
            }
        }
        currentLoad += jobs[front].cpu_load;
        result = Math.max(currentLoad, result);
        front++
    }

    return result;
}

console.log(`Maximum CPU load at any time: ${find_max_cpu_load([
    new Job(1, 4, 3), new Job(2, 5, 4), new Job(7, 9, 6),
])}`); // 7
console.log(`Maximum CPU load at any time: ${find_max_cpu_load([
    new Job(6, 7, 10), new Job(2, 4, 11), new Job(8, 12, 15),
])}`); // 15
console.log(`Maximum CPU load at any time: ${find_max_cpu_load([
    new Job(1, 4, 2), new Job(2, 4, 1), new Job(3, 6, 5),
])}`); // 8

File: merge_intervals/python/maximum-cpu-load.py
class Job:
    def __init__(self, start, end, cpu_load):
        self.start = start
        self.end = end
        self.cpu_load = cpu_load


def find_max_cpu_load(jobs):
    jobs.sort(key=lambda job: job.start)
    jobs.sort(key=lambda job: job.end)

    back = 0
    load = jobs[0].cpu_load
    max_load = load

    for front in range(1, len(jobs)):
        load += jobs[front].cpu_load
        while jobs[front].start >= jobs[back].end:
            load -= jobs[back].cpu_load
            back += 1
        max_load = max(max_load, load)

    return max_load


if __name__ == '__main__':
    print(f"Minimum meeting rooms required: {find_max_cpu_load([Job(1, 4, 3), Job(2, 5, 4), Job(7, 9, 6)])}")
    print(f"Minimum meeting rooms required: {find_max_cpu_load([Job(6, 7, 10), Job(2, 4, 11), Job(8, 12, 15)])}")
    print(f"Minimum meeting rooms required: {find_max_cpu_load([Job(1, 4, 2), Job(2, 4, 1), Job(3, 6, 5)])}")
